Moved frame to be a member variable of RigidBodySensor. Added method …

liangfok opened the pull request:
…to get the name of the model (robot) that owns the sensor.

Closes https://github.com/RobotLocomotion/drake/issues/2460

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2462) &emsp; Multiple assignees:&emsp;<img alt="@RussTedrake" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/6442292?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/RussTedrake">RussTedrake</a>,&emsp;<img alt="@jwnimmer-tri" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/17596505?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/jwnimmer-tri">jwnimmer-tri</a>,&emsp;<img alt="@liangfok" height="20" width="20" align="absmiddle" src="https://avatars.githubusercontent.com/u/1388098?s=40&v=3">&nbsp;<a href="/robotlocomotion/drake/pulls/assigned/liangfok">liangfok</a>
<!-- Reviewable:end -->

liangfok commented:
May I suggest +@jwnimmer-tri for both feature and platform review?

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> #### Moved frame to be a member variable of RigidBodySensor\. Added method …
> 
> …to get the name of the model (robot) that owns the sensor.
> 
> 
> 
> Closes https://github.com/RobotLocomotion/drake/issues/2460
> 
> 
> 
</details>


---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIifJEOWinzYKu_-NUT:-1899780784)*
<!-- Sent from Reviewable.io -->

liangfok commented:
This passed CI build https://drake-jenkins.csail.mit.edu/job/experimental/1818/

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> May I suggest +@jwnimmer-tri for both feature and platform review?
</details>


---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIilOW6ZnFRSsIANTxw:1370035413)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@liankfok Yes, I will do platform review; a bit later tonight.

If there is anyone who knows the RBS / RBF design very well, it might be worth adding them as a feature reviewer also.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This passed CI build https://drake-jenkins.csail.mit.edu/job/experimental/1818/
</details>


---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIj01h3REk--vf8U-3d:-582496161)*
<!-- Sent from Reviewable.io -->

liangfok commented:
May I suggest: +@russtedrake for feature review 

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> @liankfok Yes, I will do platform review; a bit later tonight.
> 
> If there is anyone who knows the RBS / RBF design very well, it might be worth adding them as a feature reviewer also.
</details>


---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIjE5rkvKJWtnVOXEvO:-1014310024)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.h, line 474 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjp92tTNAzl2oCZr5l:-KIjp92tTNAzl2oCZr5m:764224411) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L474)):*
> ```objc
>    * to which the sensor is attached.
>    */
>   RigidBodySensor(RigidBodySystem const& sys, const std::string& name,
> ```

BTW Consider putting the first argument's `const` in the correct place (first) since this signature is already being altered.

---

*[drake/systems/plants/RigidBodySystem.h, line 475 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjnnRiPsgJ641q83zV:-KIjnnRiPsgJ641q83zW:-1463965935) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L475)):*
> ```objc
>    */
>   RigidBodySensor(RigidBodySystem const& sys, const std::string& name,
>     const std::shared_ptr<RigidBodyFrame> frame)
> ```

Remove the `const` here (not part of the API, and I think it defeats the move constructor).

---

*[drake/systems/plants/RigidBodySystem.h, line 502 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjpLFO_uLLGuopsvKv:-KIjpLFO_uLLGuopsvKw:1054572300) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L502)):*
> ```objc
>    * The rigid body tree to which the sensor is attached.
>    */
>   RigidBodySystem const& sys_;
> ```

BTW Consider putting the `const` in the correct place (first) since this line is already being altered.

---

*[drake/systems/plants/RigidBodySystem.h, line 506 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjqFYlusFxnnG_BKRs:-KIjqFYlusFxnnG_BKRt:-1617128258) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L506)):*
> ```objc
>   /**
>    * The sensor's name.
>    */
> ```

BTW Minor docs like this take up only 1 line when the `/// The sensor's name.` doxygen style is used.  It fits more code on our screens that way.  _Shrug._

---

*[drake/systems/plants/RigidBodySystem.h, line 507 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjpZsfVKMuz6XGB3cM:-KIjpZsfVKMuz6XGB3cN:529182587) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L507)):*
> ```objc
>    * The sensor's name.
>    */
>   std::string name_;
> ```

BTW Consider marking this const.

---

*[drake/systems/plants/RigidBodySystem.h, line 512 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjpjgVEd7TG64flOe1:-KIjpjgVEd7TG64flOe2:690772096) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L512)):*
> ```objc
>    * The frame within the rigid body tree to which this sensor is attached.
>    */
>   const std::shared_ptr<RigidBodyFrame> frame_;
> ```

Protected fields are verboten in cppguide; a protected accessor may be offered instead.
https://google.github.io/styleguide/cppguide.html#Access_Control

(Note that once private, the field itself must not have any doxygen-styled comments.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@RussTedrake The only question for you is whether every `RigidBodySensor` will always have a `RigidBodyFrame` associated with it.  So far, every subclass I can find in-tree has had a frame; will that always be true?

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> May I suggest: +@russtedrake for feature review 
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIjxjrI2ncJyu0jorne:-1636028487)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
yes -- i think that will be true for most sensors and is reasonable to impose.  one example where it was not true (in matlab) is the "FullStateSensor" which simply passes through the full state, but that one is an anomaly.    

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> @RussTedrake The only question for you is whether every `RigidBodySensor` will always have a `RigidBodyFrame` associated with it.  So far, every subclass I can find in-tree has had a frame; will that always be true?
</details>


---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIlYByNIBIcObv1pgl9:-1956561995)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Interesting. For the "FullStateSensor" I would simply attach it to a sensor frame that is equivalent (identity transform) to the world frame.

<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> yes -- i think that will be true for most sensors and is reasonable to impose.  one example where it was not true (in matlab) is the "FullStateSensor" which simply passes through the full state, but that one is an anomaly.    
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIlgOC1IudJCl5QA3-n:-1086199249)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/systems/plants/RigidBodySystem.h, line 506 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjqFYlusFxnnG_BKRs:-KImYUDA1JIlttcMJRTD:-1865010298) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L506)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Minor docs like this take up only 1 line when the `/// The sensor's name.` doxygen style is used.  It fits more code on our screens that way.  _Shrug._
</details>
Done. I went ahead and modified all of the one-line documentation blocks to use `///`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Great, thanks both.  I am convinced that this PR's design change is reasonable.   A few diff discussions remain open, and then I can merge.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Interesting. For the "FullStateSensor" I would simply attach it to a sensor frame that is equivalent (identity transform) to the world frame.
</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIm_jzj9nC9UlVps_q6:-933088291)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I've addressed your review comments in my local branch (thanks for making them)  but am waiting for my computer to compile before I push the code.  It'll probably have to wait until after the various lunch / afternoon meetings.

<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Great, thanks both.  I am convinced that this PR's design change is reasonable.   A few diff discussions remain open, and then I can merge.
</details>


---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/RigidBodySystem.h, line 474 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjp92tTNAzl2oCZr5l:-KImY-KMxxFOwseOpSrw:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L474)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider putting the first argument's `const` in the correct place (first) since this signature is already being altered.
</details>
Done.

---

*[drake/systems/plants/RigidBodySystem.h, line 475 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjnnRiPsgJ641q83zV:-KImY63nNmsW_3IbCXeg:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L475)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Remove the `const` here (not part of the API, and I think it defeats the move constructor).
</details>
Done.

---

*[drake/systems/plants/RigidBodySystem.h, line 502 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjpLFO_uLLGuopsvKv:-KImYCIlE1jzjksaPx-Y:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L502)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider putting the `const` in the correct place (first) since this line is already being altered.
</details>
Done.

---

*[drake/systems/plants/RigidBodySystem.h, line 507 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjpZsfVKMuz6XGB3cM:-KImZDb8usTzdZQLJqVI:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L507)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider marking this const.
</details>
Done.

---

*[drake/systems/plants/RigidBodySystem.h, line 512 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIjpjgVEd7TG64flOe1:-KImZ_zAEFOzQvsaP4Xg:1707459498) ([raw file](https://github.com/robotlocomotion/drake/blob/b4cd75bda1ee613cf5f703baf0ca70b1fd693ae8/drake/systems/plants/RigidBodySystem.h#L512)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Protected fields are verboten in cppguide; a protected accessor may be offered instead.
> https://google.github.io/styleguide/cppguide.html#Access_Control
> 
> (Note that once private, the field itself must not have any doxygen-styled comments.)
</details>
Done. I went ahead and made the newly added accessor, `get_frame()`, public. I figured there's no reason to keep it protected and it may be useful for some other object in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KImaODFKz0OVTfkfW_5:-2022586510)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/plants/RigidBodySystem.h, line 495 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIo1H2Ri9BI_68bJ2Zl:-KIo1H2Ri9BI_68bJ2Zm:901604328) ([raw file](https://github.com/robotlocomotion/drake/blob/ca5f60a753431d24acc63aa5125186f332667f45/drake/systems/plants/RigidBodySystem.h#L495)):*
> ```objc
> 
>   /// Returns the frame to which thi sensor is attached.
>   const std::shared_ptr<RigidBodyFrame> get_frame() const;
> ```

Sorry for all the const trouble, but the first one should not appear here; the const qualifier on a return-by-value is not relevant (and so is misleading).

However... that's likely irrelevant anyway...

Returning by shared_ptr is only necessary for calling code that wants to share ownership; is that the case here?  If the only intent is that calling code wants to know the Frame's values (and not extend its lifetime), the return type should be `const RigidBodyFrame&`.  Returning as a shared_ptr brings into question whether the reference count of the Frame is changing.  It looks like this is not intended, and that return-by-const-ref is the best choice here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/plants/RigidBodySystem.h, line 495 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-KIo1H2Ri9BI_68bJ2Zl:-KIo9Nc96be6Uhmukwgr:122093682) ([raw file](https://github.com/robotlocomotion/drake/blob/ca5f60a753431d24acc63aa5125186f332667f45/drake/systems/plants/RigidBodySystem.h#L495)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Sorry for all the const trouble, but the first one should not appear here; the const qualifier on a return-by-value is not relevant (and so is misleading).
> 
> However... that's likely irrelevant anyway...
> 
> Returning by shared_ptr is only necessary for calling code that wants to share ownership; is that the case here?  If the only intent is that calling code wants to know the Frame's values (and not extend its lifetime), the return type should be `const RigidBodyFrame&`.  Returning as a shared_ptr brings into question whether the reference count of the Frame is changing.  It looks like this is not intended, and that return-by-const-ref is the best choice here.
</details>
Done. Thanks for the review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I've addressed your review comments in my local branch (thanks for making them)  but am waiting for my computer to compile before I push the code.  It'll probably have to wait until after the various lunch / afternoon meetings.
</details>


---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462#-:-KIoLVQPBFNbzMigRBq2:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2462)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Using override-merge due to out-of-date-with-master.
